app/src/test/java/com/training/trainingapp/view_model/SignUpViewModelTest.kt

```diff
@@ -0,0 +1,117 @@
+import org.junit.Before
+import org.junit.Rule
+import org.junit.Test
+
+@OptIn(ExperimentalCoroutinesApi::class, DelicateCoroutinesApi::class)
+class SignUpViewModelTest {
+
+    @ExperimentalCoroutinesApi
+    @get:Rule
+    var mainCoroutineRule = MainCoroutineRule()
+
+    lateinit var sampleViewModel: SignUpViewModel
+
+    val signUpViewUseCase = mockk<SignUpViewUseCase>()
+
+    @Before
+    fun initValues() {
+        sampleViewModel = SignUpViewModel(signUpViewUseCase)
+    }
+
+    @Test
+    fun givenEmail_WhenEmailChanged_ThenViewModelEmailMustBeChanged() {
+        var email = "paria.m7616@gmial.com"
+        sampleViewModel.onEmailChanged(email)
+        assertEquals(sampleViewModel.pageState.value.email, email)
+
+        email = "paria"
+        sampleViewModel.onEmailChanged(email)
+        assertEquals(sampleViewModel.pageState.value.email, email)
+
+        email = ""
+        sampleViewModel.onEmailChanged(email)
+        assertEquals(sampleViewModel.pageState.value.email, email)
+    }
+
+    @Test
+    fun givenCorrectEmail_WhenEmailChangedAndSubmitClicked_ThenViewModelEmailValidateShouldBeTrue() {
+        coEvery { signUpViewUseCase.invoke(any()) } returns SignUpResponse(true)
+
+        val email = "paria.m7616@gmial.com"
+        sampleViewModel.onEmailChanged(email)
+        sampleViewModel.onSubmitButtonClicked()
+        assertTrue(sampleViewModel.pageState.value.emailValidate)
+    }
+
+    @Test
+    fun givenWrongEmail_WhenEmailChangedAndSubmitClicked_ThenViewModelEmailValidateShouldBeFalse() {
+        val email = "paria."
+        sampleViewModel.onEmailChanged(email)
+        sampleViewModel.onSubmitButtonClicked()
+        assertFalse(sampleViewModel.pageState.value.emailValidate)
+    }
+
+    @Test
+    fun givenEmptyEmail_WhenEmailChangedAndSubmitClicked_ThenValidationShouldBeFalse() {
+        val email = ""
+        sampleViewModel.onEmailChanged(email)
+        sampleViewModel.onSubmitButtonClicked()
+        assertFalse(sampleViewModel.pageState.value.emailValidate)
+    }
+
+    @Test
+    fun givenCorrectEmail_WhenEmailChangedAndSubmitClicked_ThenPageStateMustBeSuccess() = runTest {
+        coEvery { signUpViewUseCase.invoke(any()) } coAnswers {
+            delay(10)
+            SignUpResponse(true)
+        }
+        val email = "paria.m7616@gmial.com"
+
+        sampleViewModel.pageState.test {
+            sampleViewModel.onEmailChanged(email)
+            assertEquals(PageState.IDLE, awaitItem().pageState)
+
+            sampleViewModel.onSubmitButtonClicked()
+            assertEquals(PageState.IDLE, awaitItem().pageState)
+            assertEquals(PageState.LOADING, awaitItem().pageState)
+            assertEquals(PageState.SUCCESS, awaitItem().pageState)
+        }
+    }
+
+    @Test
+    fun givenCorrectEmail_WhenEmailChangedAndSubmitClicked_ThenPageStateMustBeFailed() = runTest {
+        coEvery { signUpViewUseCase.invoke(any()) } coAnswers {
+            delay(10)
```

Comment:
It's better to put yeild here so that your test won't be delayed

Reply:
I'm wondering, yesterday with yield it's not working and today it's work!